### DISCUSSION TITLE
What do you mean by blackboard pattern?

### DISCUSSION THREAD

**Discussion by xmfcx** [eclipse-iceoryx/iceoryx2#485] [Q&A] [answered]
https://github.com/eclipse-iceoryx/iceoryx2/blob/5f6a0dce01eb63cfd4219c5b5054bb0a626afb9a/README.md?plain=1#L44-L46

> * Send huge amounts of data using a publish/subscribe, request/response
>   (planned), pipeline (planned) or **blackboard pattern (planned)**, making it ideal
>   for scenarios where large datasets need to be shared.

I've looked into multiple definitions of the [Blackboard Design Pattern](https://en.wikipedia.org/wiki/Blackboard_(design_pattern)) but I still cannot understand what is meant here. How is this different than the existing "publish/subscribe" messaging system?

Is this similar to what can be achieved by composable nodes? Shared memory?

**Comment by orecham** [comment, marked as answer]
@xmfcx 

The blackboard pattern is intended for scenarios where multiple participants need to collaboratively build up information in a shared data structure. i.e. The full picture cannot be built up without specific knowledge of the contributing participants.

Consider the following cases for "service discovery":

1. iceoryx2-only system
    1. Here all the information to form the full service discovery state can be obtained by iceoryx2
    2. The blackboard pattern is not needed in this case
2. iceoryx2 system extending into DDS and Zenoh
    1. Dedicated gateway processes for translating between DDS<->iceoryx2 and Zenoh<->iceoryx2 are required
    2. Only the gateway processes have information about the discovery state of their respective communication mechanisms
    3. The blackboard pattern enables these gateways and iceoryx2 processes to collaboratively construct the full service discovery state, since no single component has access to all the necessary information


```mermaid
flowchart TB
    subgraph "Service Discovery"
        subgraph "iceoryx2 Domain"
            A2[iceoryx2 Process A]
            B2[iceoryx2 Process B]
            DG[DDS Gateway]
            ZG[Zenoh Gateway]
        end
        
        BB((Service Discovery Blackboard)) <--> A2
        BB <--> B2
        
        subgraph "DDS Domain"
            DE[DDS Endpoint]
        end

        subgraph "Zenoh Domain"
            ZE[Zenoh Endpoint]
        end
        
        BB <--> DG
        BB <--> ZG
        
        DG <--> |"DDS"| DE
        ZG <--> |"Zenoh"| ZE
        
        style BB fill:#FF8C00,stroke:#333,stroke-width:4px
    end
```

**Reply by xmfcx** [reply]
Wow thank you so much for the quick and clear explanation! It clears it up a lot for me.
